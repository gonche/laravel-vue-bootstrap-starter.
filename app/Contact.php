<?php

namespace App;

use Carbon\Carbon;
use Illuminate\Database\Eloquent\Model;

class Contact extends Model
{
    /**
     * The attributes that are mass assignable.
     *
     * @var array
     */
    protected $fillable = [
        'name', 'company', 'email', 'phone', 'message',
    ];

    public function getCreatedAtAttribute(string $createdAt): string
    {
        return Carbon::parse($createdAt)->format('F jS, Y h:i:s A');
    }
}
