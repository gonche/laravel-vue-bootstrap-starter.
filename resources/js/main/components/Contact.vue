<template>
    <div class="mt-10">
        <h5>Example Component for &lt;Contact&gt;</h5>
        <form
            method="post"
            novalidate
            @submit.prevent="submit"
        >
            <!-- Name -->
            <div class="form-group">
                <label for="name">Name</label>
                <input
                    id="name"
                    v-model="$v.formData.name.$model"
                    type="text"
                    class="form-control"
                    :class="{ 'is-invalid': attemptSubmit && $v.formData.name.$invalid, 'is-valid': attemptSubmit && !$v.formData.name.$invalid }"
                    placeholder="Enter name"
                >
                <div
                    v-if="!$v.formData.name.required && attemptSubmit"
                    class="invalid-feedback"
                >
                    This field is required
                </div>
                <div
                    v-if="!$v.formData.name.maxLength && attemptSubmit"
                    class="invalid-feedback"
                >
                    This maximum is 255
                </div>
                <div
                    v-if="!$v.formData.name.minLength && attemptSubmit"
                    class="invalid-feedback"
                >
                    This minimum length is 2
                </div>
            </div>

            <!-- Company -->
            <div class="form-group">
                <label for="company">Company</label>
                <input
                    id="company"
                    v-model="$v.formData.company.$model"
                    type="text"
                    class="form-control"
                    :class="{ 'is-invalid': attemptSubmit && $v.formData.company.$invalid, 'is-valid': attemptSubmit && !$v.formData.company.$invalid }"
                    placeholder="Enter company"
                >
                <div
                    v-if="!$v.formData.company.required && attemptSubmit"
                    class="invalid-feedback"
                >
                    This field is required
                </div>
                <div
                    v-if="!$v.formData.company.maxLength && attemptSubmit"
                    class="invalid-feedback"
                >
                    This maximum is 255
                </div>
                <div
                    v-if="!$v.formData.company.minLength && attemptSubmit"
                    class="invalid-feedback"
                >
                    This minimum length is 2
                </div>
            </div>

            <!-- Email -->
            <div class="form-group">
                <label for="email">Email address</label>
                <input
                    id="email"
                    v-model="$v.formData.email.$model"
                    type="email"
                    class="form-control"
                    :class="{ 'is-invalid': attemptSubmit && $v.formData.email.$invalid, 'is-valid': attemptSubmit && !$v.formData.email.$invalid }"
                    aria-describedby="emailHelp"
                    placeholder="Enter email"
                >
                <div
                    v-if="!$v.formData.email.required && attemptSubmit"
                    class="invalid-feedback"
                >
                    This field is required
                </div>
                <div
                    v-if="!$v.formData.email.email && attemptSubmit"
                    class="invalid-feedback"
                >
                    Enter a valid email
                </div>

                <div
                    v-if="!$v.formData.email.maxLength && attemptSubmit"
                    class="invalid-feedback"
                >
                    This maximum is 255
                </div>
                <small
                    id="emailHelp"
                    class="form-text text-muted"
                >We'll never share your email with anyone else.</small>
            </div>

            <!-- Phone -->
            <div class="form-group">
                <label for="phone">Phone number</label>
                <input
                    id="phone"
                    v-model="$v.formData.phone.$model"
                    type="phone"
                    class="form-control"
                    :class="{ 'is-invalid': attemptSubmit && $v.formData.phone.$invalid, 'is-valid': attemptSubmit && !$v.formData.phone.$invalid }"
                    placeholder="###-###-####"
                >
                <div
                    v-if="!$v.formData.phone.required && attemptSubmit"
                    class="invalid-feedback"
                >
                    This field is required
                </div>
                <div
                    v-if="!$v.formData.phone.maxLength && attemptSubmit"
                    class="invalid-feedback"
                >
                    This maximum is 255
                </div>
            </div>

            <!-- Message -->
            <div class="form-group">
                <label for="name">Message</label>
                <textarea
                    id="message"
                    v-model="$v.formData.message.$model"
                    class="form-control"
                    :class="{ 'is-invalid': attemptSubmit && $v.formData.message.$invalid, 'is-valid': attemptSubmit && !$v.formData.message.$invalid }"
                    placeholder="Enter a message"
                />
                <div
                    v-if="!$v.formData.message.required && attemptSubmit"
                    class="invalid-feedback"
                >
                    This field is required
                </div>
                <div
                    v-if="!$v.formData.message.minLength && attemptSubmit"
                    class="invalid-feedback"
                >
                    Please tell us more
                </div>
            </div>

            <!-- Button -->
            <button
                id="submit-button"
                type="submit"
                class="btn btn-primary"
                :disabled="submitStatus === 'PENDING'"
            >
                Submit
            </button>
            &nbsp;<a
                href="javascript:void(0)"
                @click="clearForm"
            >Clear</a>
            <p
                v-if="submitStatus === 'OK'"
                class="alert alert-success mt-2"
            >
                Thanks for your submission!
            </p>
            <p
                v-if="submitStatus === 'ERROR'"
                class="alert alert-warning mt-2"
            >
                Please fill the form correctly.
            </p>
            <p
                v-if="submitStatus === 'PENDING'"
                class="alert alert-info mt-2"
            >
                Sending...
            </p>
            <p
                v-if="submitStatus === 'FAILED'"
                class="alert alert-danger mt-2"
            >
                Failed to complete your request
            </p>
        </form>
    </div>
</template>
<script>
import { required, email, maxLength, minLength } from 'vuelidate/lib/validators'
import axios from 'axios'

export default {
  data: function () {
    return {
      errorMessage: null,
      formData: {
        email: '',
        phone: '',
        name: '',
        company: '',
        message: ''
      },
      attemptSubmit: false,
      submitStatus: null
    }
  },
  validations: {
    formData: {
      email: { required, email, maxLength: maxLength(255) },
      phone: { required, maxLength: maxLength(255) },
      name: { required, maxLength: maxLength(255), minLength: minLength(2) },
      company: { required, maxLength: maxLength(255), minLength: minLength(2) },
      message: { required, minLength: minLength(10) }
    }
  },
  methods: {
    async makeRequest () {
      try {
        const vm = this
        // send a POST request
        return axios.post(
          '/api/contacts',
          {
            name: this.formData.name,
            email: this.formData.email,
            phone: this.formData.phone,
            company: this.formData.company,
            message: this.formData.message
          }
        )
          .then(function (response) {
            // handle success
            vm.clearForm()
            vm.submitStatus = 'OK'
            return response
          })
          .catch(function (error) {
            // handle error
            vm.errorMessage = 'Request failed'
            vm.submitStatus = 'FAILED'
            return error
          })
          // .then(function () {
          // always executed
          //  return
          // })
      } catch (ex) {
        console.log(ex)
      }
    },
    submit (e) {
      this.attemptSubmit = true

      this.$v.$touch()

      if (this.$v.$invalid) {
        this.submitStatus = 'ERROR'
      } else {
        // do your submit logic here
        this.submitStatus = 'PENDING'
        this.makeRequest()
      }
    },
    clearForm: function () {
      this.$v.$reset()
      Object.assign(this.$data, this.$options.data.apply(this))
    }
  }
}
</script>
