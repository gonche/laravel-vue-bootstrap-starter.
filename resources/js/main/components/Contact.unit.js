// resources/js/main/components/Contact.vue
import Contact from './Contact.vue'
import axios from 'axios'

jest.mock('axios') // axios here is the mock from above!

describe('Contact.vue Component', () => {
  let wrapper = null

  const formData = {
    name: 'John Doe',
    email: 'test@test.com',
    phone: '555-555-5555',
    company: 'test',
    message: 'message message'
  }

  // creates a new wrapper after each test
  beforeEach(() => {
    wrapper = mount(Contact, {
      ...createComponentMocks()
    })
  })

  // destroy the wrapper after each test
  afterEach(() => {
    wrapper.destroy()
  })

  test('exports a valid component', () => {
    expect(Contact).toBeAComponent()
  })

  test('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('clearForm() reset data', () => {
    wrapper.setData({
      formData
    })
    wrapper.vm.clearForm()
    expect(wrapper.vm.formData.email).toBe('')
  })

  test('the output after click contains the errors', () => {
    const form = wrapper.find('form')

    form.trigger('submit.prevent')

    wrapper.vm.$forceUpdate()

    wrapper.vm.$nextTick().then(() => {
      expect(wrapper.html()).toContain('field is required')
    })
    expect(wrapper.vm.submitStatus).toBe('ERROR')
    expect(wrapper.vm.attemptSubmit).toBe(true)
  })

  test('the output after successful sumbit', async () => {
    axios.post.mockResolvedValue(Promise.resolve({ data: 'some data' }))
    wrapper.setData({
      formData
    })

    const form = wrapper.find('form')

    form.trigger('submit.prevent')

    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.submitStatus).toBe('PENDING')
      expect(wrapper.vm.attemptSubmit).toBe(true)
      expect(axios.post).toBeCalledWith('/api/contacts', formData)
    })
  })

  test('a successful request in makeRequest()', async () => {
    axios.post.mockResolvedValue(Promise.resolve({ data: 'some data' }))

    // return a promise because of the async call
    return wrapper.vm.makeRequest().then(() => {
      expect(wrapper.vm.submitStatus).toBe('OK')
      expect(wrapper.vm.formData.name).toBe('')
      expect(wrapper.vm.formData.email).toBe('')
      expect(wrapper.vm.formData.phone).toBe('')
      expect(wrapper.vm.formData.company).toBe('')
      expect(wrapper.vm.formData.message).toBe('')

      expect(wrapper.html()).toContain('Thanks for your submission!')
    })
  })

  test('an unsuccessful request in makeRequest()', async () => {
    axios.post.mockReturnValue(Promise.reject(new Error('This is an error in the request')))

    wrapper.setData({
      formData
    })
    // return a promise because of the async call
    return wrapper.vm.makeRequest().then(() => {
      expect(wrapper.vm.submitStatus).toBe('FAILED')
      expect(axios.post).toBeCalledWith('/api/contacts', formData)
      expect(wrapper.vm.formData.name).toBe(formData.name)
      expect(wrapper.vm.formData.email).toBe(formData.email)
    })
  })
})
