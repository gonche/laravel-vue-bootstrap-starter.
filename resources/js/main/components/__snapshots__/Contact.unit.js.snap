// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact.vue Component matches snapshot 1`] = `
<div class="mt-10">
  <h5>Example Component for &lt;Contact&gt;</h5>
  <form method="post" novalidate="novalidate">
    <div class="form-group"><label for="name">Name</label> <input id="name" type="text" placeholder="Enter name" class="form-control">
      <!---->
      <!---->
      <!---->
    </div>
    <div class="form-group"><label for="company">Company</label> <input id="company" type="text" placeholder="Enter company" class="form-control">
      <!---->
      <!---->
      <!---->
    </div>
    <div class="form-group"><label for="email">Email address</label> <input id="email" type="email" aria-describedby="emailHelp" placeholder="Enter email" class="form-control">
      <!---->
      <!---->
      <!----> <small id="emailHelp" class="form-text text-muted">We'll never share your email with anyone else.</small>
    </div>
    <div class="form-group"><label for="phone">Phone number</label> <input id="phone" type="phone" placeholder="###-###-####" class="form-control">
      <!---->
      <!---->
    </div>
    <div class="form-group"><label for="name">Message</label> <textarea id="message" placeholder="Enter a message" class="form-control"></textarea>
      <!---->
      <!---->
    </div> <button id="submit-button" type="submit" class="btn btn-primary">
      Submit
    </button>
    &nbsp;<a href="javascript:void(0)">Clear</a>
    <!---->
    <!---->
    <!---->
    <!---->
  </form>
</div>
`;
